metron-platform/metron-test-utilities/src/main/java/org/apache/metron/test/mock/MockHTable.java

```diff
@@ -437,15 +463,14 @@ public ResultScanner getScanner(byte[] family, byte[] qualifier)
     return getScanner(scan);
   }
 
-  List<Put> putLog = new ArrayList<>();
```

Comment:
Couldn't you just change this to
```List<Put> putLog = Collections. synchronizedList(new ArrayList<>());```?

I'd still keep the `ImmutableList.copyOf()` in `getPutLog()`, because a synchronized list still requires a synchronized block during iteration and callers could do whatever.  Notably, `ImmutableList.copyOf()` specifically states "This method is safe to use even when elements is a synchronized or concurrent collection that is currently being modified by another thread."

At that point, it should be thread-safe.  It doesn't expose the underlying list, the putLog doesn't get iterated over in this class, and it only exposes thread safe copies of the putLog.

I'm also not convinced that the Supplier solution is actually thread-safe, which may be my own ignorance. The `get()` itself is thread-safe (per `memoize()`), but just returns a regular ArrayList (```this.putLog = Suppliers.memoize(() -> new ArrayList<Put>());```).  The `add()` won't necessarily be thread-safe, because two threads could each call `get()` and still attempt to put onto the same list at the same time. Assuming that's correct (which I'd like a second opinion on), it's probably just more that shuffling things happens to ease the race condition.

Somebody feel free to hop in and correct any misconceptions I have here.  I'm not as familiar with Java's concurrency as I probably (definitely?) should be.